File: frontend/src/services/preview.ts
/**
 * Preview Mode Service
 *
 * Story 1.13: Bot Preview Mode
 *
 * Provides client-side API for preview mode functionality:
 * - Start a preview session
 * - Send messages in preview mode
 * - Reset preview conversation
 *
 * API Endpoints:
 * - POST /api/v1/preview/conversation - Create a new preview session
 * - POST /api/v1/preview/message - Send a message and get bot response
 * - DELETE /api/v1/preview/conversation - Clear preview session
 */

import { apiClient } from './api';

/**
 * Starter prompts provided by the preview API
 */
export interface StarterPrompt {
  text: string;
}

/**
 * Response from POST /api/v1/preview/conversation
 */
export interface PreviewSessionResponse {
  previewSessionId: string;
  merchantId: number;
  createdAt: string;
  starterPrompts: string[];
}

/**
 * Metadata about the bot response
 */
export interface PreviewMessageMetadata {
  intent: string;
  faqMatched: boolean;
  productsFound: number;
  llmProvider: string;
}

/**
 * Product returned in bot response
 */
export interface PreviewProduct {
  product_id: string;
  title: string;
  price: number | null;
  currency: string;
  image_url: string | null;
  available: boolean;
}

/**
 * Response from POST /api/v1/preview/message
 */
export interface PreviewMessageResponse {
  response: string;
  confidence: number;
  confidenceLevel: 'high' | 'medium' | 'low';
  metadata: PreviewMessageMetadata;
  products?: PreviewProduct[];
}

/**
 * Request body for POST /api/v1/preview/message
 */
export interface PreviewMessageRequest {
  message: string;
  previewSessionId: string;
}

/**
 * Response from DELETE /api/v1/preview/conversation
 */
export interface PreviewResetResponse {
  cleared: boolean;
  message: string;
}

/**
 * MinimalEnvelope wrapper for all responses
 */
export interface PreviewEnvelope<T> {
  data: T;
  meta: {
    requestId: string;
    timestamp: string;
  };
}

/**
 * Error response from preview endpoints
 */
export interface PreviewErrorResponse {
  detail?: {
    error_code: number;
    message: string;
  };
  message?: string;
}

/**
 * Preview Mode error codes from backend
 */
export enum PreviewErrorCode {
  // Message validation errors (4300-4309)
  INVALID_MESSAGE_FORMAT = 4300,
  MESSAGE_TOO_LONG = 4301,

  // Session errors (4310-4319)
  SESSION_NOT_FOUND = 4310,
  SESSION_EXPIRED = 4350,

  // Bot response errors (4320-4329)
  FAILED_TO_GENERATE = 4303,
  BOT_CONFIG_INCOMPLETE = 4304,

  // Reset errors (4350-4359)
  FAILED_TO_RESET = 4351,
}

/**
 * Preview API Service
 */
export const previewService = {
  /**
   * Start a new preview conversation session
   *
   * POST /api/v1/preview/conversation
   *
   * @returns Promise with preview session info and starter prompts
   */
  async startPreviewSession(): Promise<PreviewSessionResponse> {
    const response = await apiClient.post<PreviewSessionResponse>(
      '/api/v1/preview/conversation'
    );
    return response.data;
  },

  /**
   * Send a message in preview mode
   *
   * POST /api/v1/preview/message
   *
   * @param message - The message text to send
   * @param previewSessionId - The preview session ID
   * @returns Promise with bot response and confidence metadata
   */
  async sendPreviewMessage(
    message: string,
    previewSessionId: string
  ): Promise<PreviewMessageResponse> {
    const request: PreviewMessageRequest = {
      message,
      previewSessionId,
    };

    const response = await apiClient.post<PreviewMessageResponse>(
      '/api/v1/preview/message',
      request
    );
    return response.data;
  },

  /**
   * Reset the current preview conversation
   *
   * DELETE /api/v1/preview/conversation/{preview_session_id}
   *
   * @param previewSessionId - The preview session ID to reset
   * @returns Promise with reset confirmation
   */
  async resetPreviewConversation(
    previewSessionId: string
  ): Promise<PreviewResetResponse> {
    const response = await apiClient.delete<PreviewResetResponse>(
      `/api/v1/preview/conversation/${previewSessionId}`
    );
    return response.data;
  },
};

File: backend/app/schemas/preview.py
"""Pydantic schemas for bot preview mode (Story 1.13).

Preview mode allows merchants to test their bot configuration in a sandbox
environment before exposing it to real customers.

Error codes: 4300-4399 (Bot Preview Mode)
"""

from __future__ import annotations

from typing import Any, Optional
from pydantic import Field, field_validator
from datetime import datetime

from app.schemas.base import BaseSchema, MinimalEnvelope, MetaData


# Sample conversation starters for preview mode
STARTER_PROMPTS = [
    "What products do you have under $50?",
    "What are your business hours?",
    "Show me running shoes",
    "I need help with my order",
    "Tell me about your return policy",
]


class PreviewMessageRequest(BaseSchema):
    """Request schema for sending a message in preview mode.

    Uses camelCase aliases for API compatibility with frontend.
    """

    message: str = Field(min_length=1, max_length=1000, description="Message text to send to bot")
    preview_session_id: str = Field(description="Preview session identifier")

    @field_validator("message")
    @classmethod
    def validate_message_length(cls, v: str) -> str:
        """Validate message length is within acceptable limits."""
        if len(v.strip()) == 0:
            raise ValueError("Message cannot be empty or whitespace only")
        if len(v) > 1000:
            raise ValueError("Message cannot exceed 1000 characters")
        return v.strip()


class PreviewMessageResponse(BaseSchema):
    """Response schema for bot response in preview mode.

    Includes bot response text, confidence score, and metadata
    for debugging and transparency.
    """

    response: str = Field(description="Bot's response text")
    confidence: int = Field(ge=0, le=100, description="Confidence score (0-100)")
    confidence_level: str = Field(description="Confidence level: high, medium, or low")
    metadata: PreviewMessageMetadata = Field(description="Additional metadata about the response")
    products: Optional[list[dict]] = Field(
        default=None,
        description="Products returned from search queries",
    )
    cart: Optional[dict] = Field(
        default=None,
        description="Cart state if cart operation was performed",
    )


class PreviewMessageMetadata(BaseSchema):
    """Metadata for preview message response.

    Provides technical details about the bot's response
    for debugging and transparency.
    """

    intent: Optional[str] = Field(None, description="Detected intent classification")
    faq_matched: bool = Field(description="Whether an FAQ was matched")
    products_found: int = Field(default=0, description="Number of products found in search")
    llm_provider: Optional[str] = Field(None, description="LLM provider used for response")


class PreviewSessionResponse(BaseSchema):
    """Response schema for preview session initialization.

    Returns session info and sample conversation starters.
    """

    preview_session_id: str = Field(description="Unique preview session identifier")
    merchant_id: int = Field(description="Merchant ID")
    created_at: str = Field(description="Session creation timestamp (ISO-8601)")
    starter_prompts: list[str] = Field(
        default_factory=lambda: STARTER_PROMPTS, description="Sample conversation starter prompts"
    )


class PreviewResetResponse(BaseSchema):
    """Response schema for preview session reset."""

    cleared: bool = Field(description="Whether the conversation was cleared")
    message: str = Field(description="Status message")


class PreviewMessageEnvelope(MinimalEnvelope):
    """Minimal envelope for preview message response."""

    data: PreviewMessageResponse


class PreviewSessionEnvelope(MinimalEnvelope):
    """Minimal envelope for preview session response."""

    data: PreviewSessionResponse


class PreviewResetEnvelope(MinimalEnvelope):
    """Minimal envelope for preview reset response."""

    data: PreviewResetResponse


__all__ = [
    "MinimalEnvelope",
    "MetaData",
    "PreviewMessageRequest",
    "PreviewMessageResponse",
    "PreviewMessageMetadata",
    "PreviewSessionResponse",
    "PreviewResetResponse",
    "PreviewMessageEnvelope",
    "PreviewSessionEnvelope",
    "PreviewResetEnvelope",
    "STARTER_PROMPTS",
]
